hive/agents/ppo.py

```diff
@@ -0,0 +1,413 @@
+            "advantages": np.empty(agent_traj_state["act_vars"]["value"].shape),
+        }
+        if "agent_id" in update_info:
+            preprocessed_update_info["agent_id"] = int(update_info["agent_id"])
+
+        return preprocessed_update_info
+
+    def preprocess_update_batch(self, batch):
+        """Returns preprocesed batch sampled from the replay buffer.
+
+        Args:
+            batch: Batch sampled from the replay buffer for the current update.
+        """
+        for key in batch:
+            batch[key] = torch.tensor(batch[key], device=self._device)
+
+        return batch
+
+    @torch.no_grad()
+    def get_action_logprob_value(self, observation):
+        """Returns the action, logprob, and value for the agent
+
+        Args:
+            observation: The current observation.
+        """
+        observation = torch.tensor(
+            np.expand_dims(observation, axis=0), device=self._device
+        ).float()
+        action, logprob, _, value = self._actor_critic(observation)
+        action = action.cpu().detach().numpy()
+        logprob = logprob.cpu().numpy()
+        value = value.cpu().numpy()
+        action = action[0]
+
+        return action, logprob, value
+
+    @torch.no_grad()
+    def act(self, observation, agent_traj_state=None):
+        """Returns the action for the agent.
+
+        Args:
+            observation: The current observation.
+        """
+        if agent_traj_state is None:
+            agent_traj_state = {}
+        if self._observation_normalization_fn:
+            self._observation_normalization_fn.update(observation)
+            observation = self._observation_normalization_fn(observation)
+        action, logprob, value = self.get_action_logprob_value(observation)
+        agent_traj_state["act_vars"] = dict(logprob=logprob, value=value)
+        return action, agent_traj_state
+
+    def update(self, update_info, agent_traj_state=None):
+        """
+        Updates the PPO agent.
+
+        Args:
+            update_info: dictionary containing all the necessary information to
+                update the agent. Should contain a full transition, with keys for
+                "observation", "next_observation", "action", "reward", "terminated",
+                and "truncated".
+        """
+        if not self._training:
+            return
+
+        # Add the most recent transition to the replay buffer.
+        self._replay_buffer.add(
+            **self.preprocess_update_info(update_info, agent_traj_state)
+        )
+
+        if self._replay_buffer.size() == self._transitions_per_update - 1:
+            if self._observation_normalization_fn:
+                update_info["next_observation"] = self._observation_normalization_fn(
+                    update_info["next_observation"]
+                )
+            _, _, values = self.get_action_logprob_value(
+                update_info["next_observation"]
+            )
```

Comment:
This might be more readable:
```suggestion
            if self._observation_normalization_fn:
                next_observation = self._observation_normalization_fn(
                    update_info["next_observation"]
                )
            _, _, values = self.get_action_logprob_value(
                next_observation
            )
```